src/main/java/g0301_0400/s0352_data_stream_as_disjoint_intervals/SummaryRanges.java

```diff
@@ -0,0 +1,54 @@
+package g0301_0400.s0352_data_stream_as_disjoint_intervals;
+
+// #Hard #Binary_Search #Design #Ordered_Set
+
+import java.util.*;
+
+public class SummaryRanges {
+
+    private final NavigableSet<int[]> intervals;
```

Comment:
This solution is not very fast

![image](https://user-images.githubusercontent.com/348496/145738105-7c5bd647-5446-4120-978d-688247465cd4.png)

Reply:
I found better implemetation

```java
public class Node {
    int min;
    int max;
    public Node(int min, int max) {
        this.min = min;
        this.max = max;
    }
    
    public Node(int val) {
        min = val;
        max = val;
    }
}

class SummaryRanges {
    List<Node> list;
    /** Initialize your data structure here. */
    public SummaryRanges() {
        list = new ArrayList<>();
    }
    
    public void addNum(int val) {
        int left = 0;
        int right = list.size()-1;
        int index = list.size();
        while (left <= right) {
            int mid = left + (right - left) / 2;
            Node node = list.get(mid);
            if (node.min <= val && val <= node.max) {
                return;
            } else if (val < node.min) {
                index = mid;
                right = mid - 1;
            } else if (val > node.max) {
                left = mid + 1;
            }
        }
        list.add(index, new Node(val));
    }
    
    public int[][] getIntervals() {
        for (int i = 1; i < list.size(); i++) {
            Node prev = list.get(i-1);
            Node curr = list.get(i);
            if (curr.min == prev.max+1) {
                prev.max = curr.max;
                list.remove(i--);
            }
        }
        int len = list.size();
        int[][] res = new int[len][2];
        for (int i = 0; i < len; i++) {
            Node node = list.get(i);
            res[i][0] = node.min;
            res[i][1] = node.max;
        }
        return res;
    }
}
```

Reply:
![image](https://user-images.githubusercontent.com/348496/145739230-5890287a-5c66-4afa-9353-6f32af3762bf.png)

Reply:
ok